### DISCUSSION TITLE
How to add support for Range<Time> Type

### DISCUSSION THREAD

**Discussion by booti386** [diesel-rs/diesel#3870] [Q&A] [answered]
This issue was closed, but is not solved: https://github.com/diesel-rs/diesel/issues/2055

Specifically :
> I tried to change Timerange in schema.rs to Range<Time> but can't implement HasSqlType<Range<Time>> for Pg
was not addressed.

Even with the required use in the schema, `Range<Time>` cannot be used in queries because `Range<Time>` does not implement `HasSqlType`.
The orphan rule prevents us to manually implement `HasSqlType` for `Range<Time>`, as it is done in Diesel for e.g. `Range<Date>`: https://github.com/diesel-rs/diesel/blob/v2.1.4/diesel/src/pg/types/ranges.rs#L205

So, how is it possible to use a custom `Range<T>`, such as `Range<Time>` or `Range<Float4>` in Diesel, without having to copy-paste the whole Range implementation?

**Comment by weiznich** [comment, marked as answer]
`Timerange` is not a built-in postgres range type. Diesel only supports built-in range types out of the box. For any other type you need to provide the relevant custom type implementation. That means that you essentially need to a complete custom type there. What you can do to simplify your implementation is to internally call the existing generic `FromSql`/`ToSql` implementations for `Range<T>` like this:

```rust
#[derive(FromSqlRow, AsExpression)]
#[diesel(sql_type = Timerange)]
struct MyTimeRange(Bound<chrono::NaiveTime>, Bound<chrono::NaiveTime>);

impl ToSql<TimeRange, Pg> for MyTimeRange {
    fn to_sql<'b>(
        &'b self,
        out: &mut diesel::serialize::Output<'b, '_, Pg>,
    ) -> diesel::serialize::Result {
        <(Bound<chrono::NaiveTime>, Bound<chrono::NaiveTime>)  as ToSql<Range<Time>, Pg>>::to_sql(self, out).map(Self)
    }
}

impl FromSql<TimeRange, Pg> for (MyTimeRange {
    fn from_sql(
        bytes: <Pg as diesel::backend::Backend>::RawValue<'_>,
    ) -> diesel::deserialize::Result<Self> {
        <(Bound<chrono::NaiveTime>, Bound<chrono::NaiveTime>)  as FromSql<Range<Time>, Pg>>::from_sql(bytes).map(Self)
    }
}

```

**Reply by booti386** [reply]
Okay, thank you for your answer, i didn't expect it to be so straightforward!
